master/internal/api_runs.go

```diff
@@ -239,3 +251,201 @@ func sortRuns(sortString *string, runQuery *bun.SelectQuery) error {
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	err = getQ.Scan(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	var results []*apiv1.RunActionResult
+	var visibleIDs []int32
+	var validIDs []int32
+	// associated experiments to move
+	var expMoveIds []int32
+	for _, check := range runChecks {
+		visibleIDs = append(visibleIDs, check.ID)
+		if check.Archived {
+			results = append(results, &apiv1.RunActionResult{
+				Error: "Run is archived.",
+				Id:    check.ID,
+			})
+		} else {
+			if check.IsMultitrial && req.CloneMultitrial {
+				err := cloneExperimentAndRun(ctx, *check.ExpID, check.ID, req.DestinationProjectId)
+				if err != nil {
+					results = append(results, &apiv1.RunActionResult{
+						Error: fmt.Sprintf(`Failed to clone multi-trial run: %s`, err),
+						Id:    check.ID,
+					})
+				} else {
+					results = append(results, &apiv1.RunActionResult{
+						Error: "",
+						Id:    check.ID,
+					})
+				}
+				continue
+			}
+			if check.ExpID != nil {
+				expMoveIds = append(expMoveIds, *check.ExpID)
+			}
+			validIDs = append(validIDs, check.ID)
+		}
+	}
+	if req.Filter == nil {
+		for _, originalID := range req.RunIds {
+			if !slices.Contains(visibleIDs, originalID) {
+				results = append(results, &apiv1.RunActionResult{
+					Error: fmt.Sprintf("Run with id '%d' not found in project with id '%d'", originalID, req.SourceProjectId),
+					Id:    originalID,
+				})
+			}
+		}
+	}
+	if len(validIDs) > 0 {
+		tx, err := db.Bun().BeginTx(ctx, nil)
+		if err != nil {
+			return nil, err
+		}
+		defer func() {
+			txErr := tx.Rollback()
+			if txErr != nil && txErr != sql.ErrTxDone {
+				logrus.WithError(txErr).Error("error rolling back transaction in MoveRuns")
+			}
+		}()
+
+		expMoveResults, err := experiment.MoveExperiments(ctx, expMoveIds, nil, req.DestinationProjectId)
+		if err != nil {
+			return nil, err
+		}
+		failedExpMoveIds := []int32{-1}
+		for _, res := range expMoveResults {
+			if res.Error != nil {
+				failedExpMoveIds = append(failedExpMoveIds, res.ID)
+			}
+		}
+		var acceptedIDs []int32
+		if _, err = tx.NewUpdate().Table("runs").
+			Set("project_id = ?", req.DestinationProjectId).
+			Where("runs.id IN (?)", bun.In(validIDs)).
+			Where("runs.experiment_id NOT IN (?)", bun.In(failedExpMoveIds)).
+			Returning("runs.id").
+			Model(&acceptedIDs).
+			Exec(ctx); err != nil {
+			return nil, fmt.Errorf("updating run's project IDs: %w", err)
+		}
+
+		for _, acceptID := range acceptedIDs {
+			results = append(results, &apiv1.RunActionResult{
+				Error: "",
+				Id:    acceptID,
+			})
+		}
+		if err = tx.Commit(); err != nil {
+			return nil, err
+		}
+	}
+	return &apiv1.MoveRunsResponse{Results: results}, nil
+}
+
+func cloneExperimentAndRun(ctx context.Context, expID int32, runID int32, destProjID int32) error {
+	var cloneExpID int32
+
+	// clone experiment into dest project
+	err := db.Bun().NewRaw(`INSERT INTO experiments(state, config, model_definition, start_time, end_time, archived,
```

Comment:
can we use a transaction here and the insert run

Reply:
updated

Reply:
updated